packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts

```diff
@@ -68,6 +71,48 @@ export function source() {
     }
   }
 
+  function isTypedArray(obj: any, constructor: Function): boolean {
+    try {
+      return obj instanceof constructor || Object.prototype.toString.call(obj) === `[object ${constructor.name}]`;
+    } catch (error) {
+      return false;
+    }
+  }
+
+  const typedArrayConstructors: Record<TypedArrayKind, Function> = {
+    i8: Int8Array,
+    ui8: Uint8Array,
+    ui8c: Uint8ClampedArray,
+    i16: Int16Array,
+    ui16: Uint16Array,
+    i32: Int32Array,
+    ui32: Uint32Array,
+    // TODO: add Float16Array once it's in baseline
+    f32: Float32Array,
+    f64: Float64Array,
+    bi64: BigInt64Array,
+    bui64: BigUint64Array,
+  };
+
+  function typedArrayToBase64(array: any) {
+    /**
+     * Firefox does not support iterating over typed arrays, so we use `.toBase64`.
+     * Error: 'Accessing TypedArray data over Xrays is slow, and forbidden in order to encourage performant code. To copy TypedArrays across origin boundaries, consider using Components.utils.cloneInto().'
+     */
+    if ('toBase64' in array)
+      return array.toBase64();
+    const binary = Array.from(new Uint8Array(array.buffer)).map(b => String.fromCharCode(b)).join('');
```

Comment:
It is incorrect to serialize `array.buffer` without looking at `array.byteOffset` and `array.byteLength`, because underlying buffer may be longer.

Reply:
good catch! fixed in https://github.com/microsoft/playwright/pull/34949/commits/769fda67ee0c5f460b8a45d4f4b49dec17b3936b